wagtail/snippets/views/snippets.py

```diff
@@ -91,31 +92,38 @@ def dispatch(self, request, *args, **kwargs):
         return super().dispatch(request, *args, **kwargs)
 
     def get_context_data(self, **kwargs):
-        snippet_model_opts = sorted(
-            self.snippet_types, key=lambda x: x.verbose_name.lower()
+        snippet_types = sorted(
+            self.snippet_types, key=lambda x: x["model_opts"].verbose_name.lower()
         )
-        return super().get_context_data(snippet_model_opts=snippet_model_opts, **kwargs)
+        return super().get_context_data(snippet_types=snippet_types, **kwargs)
 
 
 class List(IndexView):
+    viewset = None
     any_permission_required = ["add", "change", "delete"]
     paginate_by = 20
     page_kwarg = "p"
     # If true, returns just the 'results' include, for use in AJAX responses from search
     results_only = False
 
-    def setup(self, request, *args, app_label, model_name, **kwargs):
-        self.app_label = app_label
-        self.model_name = model_name
-        self.model = self._get_model()
-        self.permission_policy = ModelPermissionPolicy(self.model)
+    def setup(self, request, *args, **kwargs):
         super().setup(request, *args, **kwargs)
-
-    def _get_model(self):
-        return get_snippet_model_from_url_params(self.app_label, self.model_name)
+        self.permission_policy = ModelPermissionPolicy(self.model)
 
     def get_index_url(self):
-        return reverse("wagtailsnippets:list", args=[self.app_label, self.model_name])
+        return reverse(self.viewset.get_url_name("list"))
+
+    def get_index_results_url(self):
+        return reverse(self.viewset.get_url_name("list_results"))
+
+    def get_delete_multiple_url(self):
+        return reverse(self.viewset.get_url_name("delete-multiple"))
+
+    def get_add_url_name(self):
+        return self.viewset.get_url_name("add")
+
+    def get_edit_url_name(self):
+        return self.viewset.get_url_name("edit")
```

Comment:
Instead of having these, I'm thinking about creating a template tag that allows you to call `get_url_name` of the viewset (or maybe using `get_admin_url_namespace` of the model) from the template. What do you think?